storage/repair.go

```diff
@@ -124,23 +128,280 @@ func (r shardRepairer) Repair(
+}
+
+func (h *hostSet) insertOneFromSet(oh *hostSet) {
+	for _, hbm := range *oh {
+		h.insert(hbm)
+		return
+	}
+}
+
+func (h *hostSet) contains(host topology.Host) bool {
+	if host == nil {
+		return false
+	}
+	_, ok := (*h)[host.String()]
+	return ok
+}
+
+func (h *hostSet) remove(host topology.Host) {
+	delete(*h, host.String())
+}
+
+func (h *hostSet) empty() bool {
+	return len(*h) == 0
+}
+
+func (r shardRepairer) newAdminRepairBlocksRequest(
+	pendingReplicasMap map[blockID]*hostSet,
+	idMap map[ts.Hash]ts.ID,
+	initialCapacity int64,
+) []block.ReplicaMetadata {
+	repairBlocks := make([]block.ReplicaMetadata, 0, initialCapacity)
+	for blkID, hs := range pendingReplicasMap {
+		for _, h := range *hs {
+			repairBlocks = append(repairBlocks, block.ReplicaMetadata{
+				ID:   idMap[blkID.idHash],
+				Host: h.Host,
+				Metadata: block.Metadata{
+					Start:    blkID.start,
+					Checksum: h.Checksum,
+					Size:     h.Size,
+				},
+			})
+		}
+	}
+	return repairBlocks
+}
+
+// newPendingReplicasMap returns a map from (series_id, block_start) -> set of host replicas,
+// using which repairs are to be performed
+//
+// NB(prateek): In the scenario when multiple peers have identical replicas for a single block, the current
+// implementation arbitrarily picks one of the peers as the source to request the replica from. In the event
+// the replica is unavailable from the selected peer, the current implementation fails. If this becomes an issue,
+// the code should enhanced to provide a list of potential sources for a single replica (as opposed to the
+// non-deterministic selection)
+func (r shardRepairer) newPendingReplicasMap(
+	diffRes repair.MetadataComparisonResult,
+	originHost topology.Host,
+) (map[blockID]*hostSet, map[ts.Hash]ts.ID) {
+	// TODO(prateek): pooling object creation in this method
+	pendingReplicas := make(map[blockID]*hostSet, diffRes.NumBlocks)
+	idMap := make(map[ts.Hash]ts.ID, diffRes.NumBlocks)
+
+	// add size differences
+	for _, rsm := range diffRes.SizeDifferences.Series() {
+		idHash := rsm.ID.Hash()
+		idMap[idHash] = rsm.ID
+		for start, blk := range rsm.Metadata.Blocks() {
+			blkID := blockID{idHash, start}
+			// find all unique sizes seen for this block
+			uniqueValues := make(map[int64]*hostSet, 3)
+			for _, hBlk := range blk.Metadata() {
+				sz := hBlk.Size
+				if _, ok := uniqueValues[sz]; !ok {
+					uniqueValues[sz] = newHostSet()
+				}
+				uniqueValues[sz].insert(hBlk)
+			}
+			for _, hs := range uniqueValues {
+				if hs.contains(originHost) {
+					// we already have originHost data available locally, no need to fetch it
+					continue
+				}
+				// insert value into replicaState
+				if _, ok := pendingReplicas[blkID]; !ok {
+					pendingReplicas[blkID] = newHostSet()
+				}
+				// only care about a single value from the unique set
+				pendingReplicas[blkID].insertOneFromSet(hs)
+			}
+		}
+	}
+
+	// add checksum differences
+	for _, rsm := range diffRes.ChecksumDifferences.Series() {
```

Comment:
The logic here looks rather similar to the one above, any chance for refactoring?